File: backend/app/routes/paper.py
from core.context import AppContext, get_ctx, active_context
"""模拟盘（Paper Trading）路由：虚拟资金撮合 + 真实行情盯市。

引擎实例由集成方注入 `state.paper_engine`（未注入时统一返回 503）。
"""
import logging

from fastapi import APIRouter, Depends

from app.routes._common import err, ok

log = logging.getLogger("qmt_work.routes.paper")

router = APIRouter()


def _engine():
    """取模拟盘引擎；未初始化返回 None（active_context() 可能尚未挂载该属性）。"""
    return getattr(active_context(), "paper_engine", None)


_NOT_READY = "模拟盘引擎未初始化"


@router.post("/paper/reset")
async def paper_reset(body: dict | None = None, ctx: AppContext = Depends(get_ctx)):
    """重置模拟盘：清空持仓/成交并重设初始资金。body: {initial_capital?}"""
    e = _engine()
    if e is None:
        return err(503, _NOT_READY)
    body = body or {}
    try:
        initial = float(body.get("initial_capital") or 1_000_000.0)
        return ok(e.reset(initial))
    except (TypeError, ValueError) as exc:
        return err(400, str(exc))


@router.post("/paper/order")
async def paper_order(body: dict, ctx: AppContext = Depends(get_ctx)):
    """模拟下单（立即以给定价格成交）。body: {code, side, price, volume, price_type?, remark?}"""
    e = _engine()
    if e is None:
        return err(503, _NOT_READY)
    body = body or {}
    try:
        order = e.submit_order(
            code=body.get("code", ""),
            side=body.get("side", ""),
            price=body.get("price", 0),
            volume=body.get("volume", 0),
            price_type=body.get("price_type", "limit"),
            remark=body.get("remark", ""),
        )
    except (TypeError, ValueError) as exc:
        return err(400, str(exc))
    db = getattr(ctx, "db", None)
    if db is not None:
        try:
            db.audit("paper", "paper.order", order["code"],
                     {"side": order["side"], "price": order["price"],
                      "volume": order["volume"]}, "filled")
        except (AttributeError, OSError, RuntimeError) as exc:
            # 审计失败不影响模拟撮合：仅丢一条审计记录
            log.debug("paper 审计写入失败（已忽略）：%s", exc)
    return ok(order)


@router.get("/paper/account")
async def paper_account(ctx: AppContext = Depends(get_ctx)):
    """模拟盘资产：现金 / 市值 / 总资产 / 浮动与已实现盈亏。

    市值与浮动盈亏基于**实时行情**最新价盯市（来自同步引擎行情缓存，不编造价格）。
    """
    e = _engine()
    if e is None:
        return err(503, _NOT_READY)
    se = getattr(ctx, "sync_engine", None)
    if se is not None:
        # 从行情缓存抽取最新价（兼容 last/lastPrice/price/close 字段）
        price_map = {}
        for code, q in (getattr(se, "latest_quotes", {}) or {}).items():
            if not isinstance(q, dict):
                continue
            for k in ("last", "lastPrice", "price", "close"):
                v = q.get(k)
                if v:
                    try:
                        price_map[code] = float(v)
                        break
                    except (TypeError, ValueError):
                        continue
        e.sync_from_map(price_map)
    return ok(e.get_account())


@router.get("/paper/positions")
async def paper_positions(ctx: AppContext = Depends(get_ctx)):
    """获取paper / positions（GET /paper/positions）。"""
    e = _engine()
    if e is None:
        return err(503, _NOT_READY)
    return ok(e.get_positions())


@router.get("/paper/trades")
async def paper_trades(limit: int = 50, ctx: AppContext = Depends(get_ctx)):
    """获取paper / trades（GET /paper/trades）。"""
    e = _engine()
    if e is None:
        return err(503, _NOT_READY)
    return ok(e.get_trades(limit))


@router.get("/paper/metrics")
async def paper_metrics(ctx: AppContext = Depends(get_ctx)):
    """获取paper / metrics（GET /paper/metrics）。"""
    e = _engine()
    if e is None:
        return err(503, _NOT_READY)
    return ok(e.metrics())

File: frontend-next/src/services/api/paper.ts
import { http } from "../http";
import type { PriceType, Side } from "@/shared/types";

/**
 * 模拟盘（Paper Trading）API 层。
 *
 * ★ 口径来源：backend/app/routes/paper.py + engines/paper_engine.py（逐字段核对）。
 *   - POST /paper/reset      {initial_capital?} → 返回**重置后的 account**（不是 ok 标志）
 *   - POST /paper/order      {code, side, price, volume, price_type?, remark?}
 *   - GET  /paper/account    现金/市值/总资产/浮动与已实现盈亏（含 positions 数组）
 *   - GET  /paper/positions  持仓列表
 *   - GET  /paper/trades?limit=50  → 元素为 {id, code, side, price, volume, ts, pnl}
 *   - GET  /paper/metrics    绩效摘要（胜率/均值/最好最差）
 *
 * 零 mock：模拟盘引擎未初始化时后端返回 503，http.ts 原样抛 ApiError，本层不粉饰。
 * 页面据此显式提示「引擎未就绪」，而不是展示一张空表冒充「没有成交」。
 */

/** 单条持仓（paper_engine.get_positions）。 */
export interface PaperPosition {
  code: string;
  name: string;
  volume: number;
  avg_cost: number;
  last_price: number;
  side?: string;
  market_value: number;
  cost: number;
  unrealized_pnl?: number;
  /** 盯市来源：live=真实行情价；cost=退化成本价（无行情） */
  marking?: "live" | "cost";
}

/** 账户概览（paper_engine.get_account，含内嵌 positions）。 */
export interface PaperAccount {
  cash: number;
  market_value: number;
  total_assets: number;
  initial: number;
  unrealized_pnl: number;
  realized_pnl: number;
  /** 小数形态（0.0123 = 1.23%），展示需 ×100 */
  total_return: number;
  position_count: number;
  /** live=任一持仓有真实行情；frozen=全部退化为成本价 */
  marking_source: "live" | "frozen";
  positions: PaperPosition[];
}

/** 单条成交（paper_trades 表原始列，注意主键是 id 不是 trade_id）。 */
export interface PaperTrade {
  id: number;
  code: string;
  side: Side;
  price: number;
  volume: number;
  ts: string;
  /** 仅卖出平仓有值；买入建仓为 0/null */
  pnl?: number | null;
}

/** 绩效摘要（paper_engine.metrics）。 */
export interface PaperMetrics {
  trade_count: number;
  close_count: number;
  win_count: number;
  win_rate: number;
  avg_pnl: number;
  best_pnl: number;
  worst_pnl: number;
  realized_pnl: number;
  unrealized_pnl: number;
  total_pnl: number;
}

export interface PaperOrderInput {
  code: string;
  side: Side;
  price: number;
  volume: number;
  price_type?: PriceType;
  remark?: string;
}

export const paperApi = {
  account: () => http.get<PaperAccount>("/paper/account"),

  positions: () => http.get<PaperPosition[]>("/paper/positions"),

  trades: (limit = 50) =>
    http.get<PaperTrade[]>("/paper/trades", { query: { limit } }),

  metrics: () => http.get<PaperMetrics>("/paper/metrics"),

  /** 模拟下单：立即以给定价格全额成交；现金/可卖不足时后端抛 400。 */
  order: (body: PaperOrderInput) => http.post<Record<string, unknown>>("/paper/order", body),

  /** 重置：清空持仓与成交、现金回到 initial_capital；返回重置后的账户。 */
  reset: (initialCapital?: number) =>
    http.post<PaperAccount>("/paper/reset", { initial_capital: initialCapital ?? 1_000_000 }),
};
